### ISSUE TITLE
make sqlite database work in browser for production

### ISSUE THREAD

**Issue by mobidev111** [typeorm/typeorm#548]
### a decent cross-platform sql implementation is available on Android, iOS & Windows with a native sqlite plugin called 'cordova-sqlite-storage'
- Cordova is a cross-platform framework covering Android, iOS & Windows
- Sqlite is available via the cordova plugin 'Cordova-sqlite-storage' https://github.com/litehelpers/Cordova-sqlite-storage - 1500+ stars & actively maintained
- it does remove most of the limitations of the native browser databases when looking from a cross-platform perspective: 
  - on iOS: removes need to use deprecated webSQL
  - on iOS & Android: features better performance than native databases
  - on Windows: Provides a working implementation

### targeting the large browser-based communities and extends the reach of typeorm massively: Angular & possibly React
 - AngularJS (Angular 1) with Ionic 1
 - Angular (Angular 2+) with Ionic 2+
 - React?

### proposal: Interface the sqlite database using the 'cordova-sqlite-storage' Cordova plugin

Possible tasks are:

- [ ] make statement: actively maintain browser support within typeorm
  - some of the existing approaches in the wild date a few months back and are not actively maintained
  - to have lasting support the browser we should maintain this as part of the typeorm/typeorm repo

- [ ] consolidate the existing browser approaches in the typeorm/typeorm repo
  - @BradyLiles could you post your excellent work in BradyLiles/ionic-orm as a pull-request against the typeorm/typeorm repo?
  - in case that @BradyLiles doesn't have time - @pleerock could you help in bringing this back into the typeorm/typeorm repo? as mentioned in https://github.com/typeorm/typeorm/issues/62#issuecomment-262148315
  - @sidneycorreia could you post your work (https://github.com/sidneycorreia/ionic-orm) as pull-request against the typeorm/typeorm repo?

- [ ] support sqlite database via cordova-sqlite-storage
  - use the sql interface -> switch the websql-based sql interface of the existing proof-of-concept to sqlite
  - adapt/create new sqlite driver, as discussed in https://github.com/typeorm/typeorm/issues/169#issuecomment-294363884

- [ ] extend example project with browser sqlite example - [typeorm/browser-example](https://github.com/typeorm/browser-example/blob/master/src/index.ts)

- [ ] make support ready for production
  - test on the major platforms Android, iOS & Windows - there are numerous issues with people interested in this: https://github.com/BradyLiles/ionic-orm/issues/5, https://github.com/typeorm/typeorm/issues/61, https://github.com/typeorm/browser-example/issues/1, https://github.com/typeorm/typeorm/issues/169
  - wipe out any of the occurring platform oddities, possibly with the help of the cordova-sqlite-storage plugin author @brodybits

- [ ] extend support into further communities
  - add support for Ionic 2+ via the Ionic native SQLite wrapper: https://ionicframework.com/docs/native/sqlite/ - @sidneycorreia could you post your pull-request (https://github.com/BradyLiles/ionic-orm/pull/6) against the typeorm/typeorm repo?
  - ...

I would love to coordinate, test-drive & give feedback to make this working in production.

Who else can contribute? Please chime in and ping possible supporters.

**Comment by pleerock** [comment]
Looks like a good plan! Currently Im quite busy with other typeorm issues and can't help much on this front, so would be great if someone take initiative on his own and implement new drivers for cordova and nativescript. This should not be hard to implement.

**Comment by daniel-lang** [comment]
Hi everybody,
I implemented a driver for cordova's sqlite plugin ([in my fork](https://github.com/daniel-lang/typeorm/tree/cordova-sqlite)) and would love to create a PR for it. But right now I'm unable to test it because of some Node.js dependencies (fs, event, stream) that make the browser version unusable.
Do you think it is possible to somehow work arround those dependencies/remove them for the browser version to work again?

**Comment by pleerock** [comment]
Thats why I created [platform](https://github.com/typeorm/typeorm/tree/master/src/platform) directory. See what `gulpfile.ts` does with it. Idea was to have all node dependencies in a single "platform class" and replace it in build process by other file which will resolve all dependencies properly.

**Comment by daniel-lang** [comment]
Okay that would mean, that every call to one of nodes modules has to go through the PlatformTools class so that `stream`, `fs` and `event` are never used directly?

**Comment by pleerock** [comment]
yeah. Everything kinda must be in "PlatformTools" file. And for browser and sqlite environment we replace this "PlatformTools" file with sqlite-specific which resolves all node dependencies its own way

**Comment by pleerock** [comment]
@daniel-lang what are your thoughts about roadmap to complete this feature request and roundup all ionic-related issues?

**Comment by daniel-lang** [comment]
I'm totally for it!
The open things from this issue are:
* Include WebSql and Cordova driver in CI
* Add NativeScript driver (#1114)

**Comment by mgleria** [comment]
Hi! I've read this thread and [#169](https://github.com/typeorm/typeorm/issues/169). Now I have to make sure that I am not completely lost and probably my questions are too basic for most of you. 
I'm beginner IONIC apps developer, and I'm trying to make an app with CRUD capabilities over 4 differents DB engines (all supported by TypeORM). When I tried to connect my app (run from the device) with MySQL remote server using mysql nodeJS module, I ran into "DriverPackageNotInstalledError: Mysql package has not been found installed" error, because I was trying to use a NodeJS module in IONIC.
I already bored you with the introduction, now I ask you:

1. Is it (or will) possible to connect to MySQL (and the others DB engines supported by TypeORM) from an Ionic App?
2. Do I have to develop some drivers in order to achieve that or 'cordova-sqlite-storage' provides me the interface that I need?
3. If so, the logic path would be: DB engine < -- > cordova-sqlite-storage < -- > TypeORM < -- > MyApp ??

Thanks for your help!!

**Comment by pleerock** [comment]
@mgleria @daniel-lang can answer better on your question, but did you try [this ionic example repository](https://github.com/typeorm/ionic-example)?

**Comment by mnewmedia** [comment]
1. It is possible
2. No driver needed, just the cordova-sqlite-plugin                                                                                     
3.1. sqliteDB <> cordova-sqlite-plugin <> typeorm <> app
3.1.mysql <> typeorm <> app

So as far as i know,  it doesnt sync the mysqlDB to the sqliteDB. But since you can use both Databases like shown (3.1 and 3.2) you can implement a synchronization by yourself.

But also better ask a maintainer about this, this is just my understanding of a quick reading through the docs.

**Comment by mgleria** [comment]
Thanks @pleerock and @mnewmedia for your replies. 

> @mgleria @daniel-lang can answer better on your question, but did you try this ionic example repository?

Yes, I'm using that example and trying to change the database connection to MySQL.

Somebody has achieve a connection with MySQL or other remote DB using TypeORM using an android device?

**Comment by daniel-lang** [comment]
1. Yes, it is possible to connect to a remote database (MySQL for example) from ionic but this functionality is currently not supported in TypeORM. I'm not sure if supporting it would be a wise idea.
2. The driver for `cordova-sqlite-storage` already exists. You can use it by setting the type in your connection options to `cordova` (as you can see in the example).
3. For your app, the path would be correct.

This means, that you will have to write a system to sync your databases between your nodejs server and the ionic app.

**Comment by mgleria** [comment]
Thanks @daniel-lang. I will work in that system to achieve de connection.

**Comment by championswimmer** [comment]
Added nativescript support here https://github.com/typeorm/typeorm/pull/2407 

review appreciated :)

**Comment by ludwiktrammer** [comment]
Does this still work with the most recent version of typeorm or had it been a regression? I can't get it to work - `getConnection` results in webpack errors and the function is undefined (see #3003).
